### DISCUSSION TITLE
ElastAlert Setup

### DISCUSSION THREAD

**Discussion by ow111** [Security-Onion-Solutions/securityonion#2519] [Unsupported Versions] [answered]
Security onion v2.3.10
Installed over network from CentOS7 , Running in docker container (docker ps -a shows all of the SO Services running OK)
This is on a Manager Node
Trying to get ElastAlert to function , for example in the *so-* index I want to filter by hostname & eventID. I currently have a rule in place in /opt/so/conf/elastalert/ which reads as the following..

_es_host: x.x.x.x
es_port: xxxx
name: myrule
type: frequency
index: *:so-*
num_events: 1
timeframe:
    hours: 4
filter:
- query:
    query_string:
      query: "winlog.event_id: xx AND agent.hostname: "xxx"
alert:
- "email"
email:
- "x@x.com"
smtp_host: xx
smtp_port: xx
from_addr: "x@x.com"
smtp_auth_file '/etc/elastalert/rules/smtp_auth_file.txt'_

I have tried to use so-elastalert-test to test my rule , but this function errors out after I give it a file - I don't know what i'm doing wrong.. 
_elastalert.util.EAException: No --config or config.yaml found , Something went wrong..._
I have also tried setting the output to debug but nothing was seen in the log file when triggering the specific eventID. I checked the elasticsearch index and the events were showing up there

I am brand new to SO, so please forgive me if this is something trivial.

**Comment by jertel** [comment, marked as answer]
You will need to upgrade to 2.3.21 in order to use the `so-elastalert-test` script.

**Reply by ow111** [reply]
Thanks!! I will get updating and try it out

**Reply by ow111** [reply]
Hi Jertel, 

I updated my Deployment and Now am seeing a different error. The script is pointing to /opt/elastalert/rules/ as default I assume you can change this. 
Inside the docker_exec Function it seems to point to the same directory of /opt/elastalert/rules , should this be /etc/elastalert/rules/ instead? I'm pretty sure my deployment using docker uses the /etc root directory.

FileNotFoundError: [Errno 2] No such file or directory: '/opt/elastalert/rules/failedlogons.yaml'

I tried creating the directory /opt/elastalert/rules and adding my file into it just to humour myself but that did not work.

**Reply by jertel** [reply]
Can you paste the full script output?

**Reply by ow111** [reply]
Traceback (most recent call last):
  File "/usr/local/bin/elastalert-test-rule", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.8/site-packages/elastalert/test_rule.py", line 445, in main
    test_instance.run_rule_test()
  File "/usr/local/lib/python3.8/site-packages/elastalert/test_rule.py", line 413, in run_rule_test
    rule_yaml = conf['rules_loader'].load_yaml(args.file)
  File "/usr/local/lib/python3.8/site-packages/elastalert/loaders.py", line 197, in load_yaml
    loaded = self.get_yaml(filename)
  File "/usr/local/lib/python3.8/site-packages/elastalert/loaders.py", line 526, in get_yaml
    return yaml_loader(filename)
  File "/usr/local/lib/python3.8/site-packages/staticconf/loader.py", line 167, in yaml_loader
    with open(filename) as fh:
FileNotFoundError: [Errno 2] No such file or directory: '/opt/elastalert/rules/failedlogons.yaml'

Test failed.

**Reply by ow111** [reply]
I can get the script to point to my rule in /opt/so/conf/elastalert/ but the docker container portion seems to be pointing to a directory which does not exist

**Reply by jertel** [reply]
All Elastalert rules must exist in the `/opt/so/rules/elastalert` directory on the host machine. Please paste the _full_ script output, including the command line you are typing, the list of rules it found, and the rule you entered at the prompt.

**Reply by ow111** [reply]
Duly Moved, The output is 
  in "/opt/elastalert/rules/failedlogons.yaml", line 12, column 7
expected <block end>, but found '<scalar>'
  in "/opt/elastalert/rules/failedlogons.yaml", line 12, column 58
pointing to the following config - 

 query:
    query_string:
      `query: "winlog.event_id: xxxx AND agent.hostname: "myhostname""`

How can I change my query to provide the desired effect?

**Comment by jertel** [comment]
Quotes within quotes should be escaped. 

`query: "some.key: \"some value\""`

**Reply by ow111** [reply]
Thanks Jertel, I will see if I can get it to work

**Reply by ow111** [reply]
I was able to get my rule to successfully run, however Now I am getting an issue back from the script saying "did not get any results".
I am pointing my rule to my elasticsearch on the manager node. When searching for the eventID 4625 in kibana I can see results. Do I have to instead point my rule to the search node?

**Reply by jertel** [reply]
Since you've overcome the original script problem I suggest creating a new discussion for tuning your Elastalert filters.
